File: 050_SmallestNumberInInfiniteSet.cpp
// Topic: Design / Heap + Set
// Smallest Number in Infinite Set: Implement a class that can pop the smallest number and add
// numbers back. Time Complexity: popSmallest O(log n) worst, addBack O(log n) Space Complexity:
// O(n)

#include <bits/stdc++.h>
using namespace std;

class SmallestInfiniteSet
{
    priority_queue<int, vector<int>, greater<int>> pq;
    unordered_set<int> inSet;
    int next;

   public:
    SmallestInfiniteSet() : next(1) {}
    int popSmallest()
    {
        if (!pq.empty())
        {
            int x = pq.top();
            pq.pop();
            inSet.erase(x);
            return x;
        }
        return next++;
    }
    void addBack(int num)
    {
        if (num < next && inSet.insert(num).second)
            pq.push(num);
    }
};

int main()
{
    SmallestInfiniteSet s;
    cout << s.popSmallest() << "\n";  // 1
    cout << s.popSmallest() << "\n";  // 2
    s.addBack(1);
    cout << s.popSmallest() << "\n";  // 1
    cout << s.popSmallest() << "\n";  // 3
    s.addBack(2);
    cout << s.popSmallest() << "\n";  // 2
    return 0;
}

File: 050_SmallestNumberInInfiniteSet.py
"""
Topic: Design / Heap
Problem: Smallest Number in Infinite Set
Description: Implement a class that returns and adds back the smallest number.
Time: push/addBack: O(log n), popSmallest: O(log n)
Space: O(n)
"""

import heapq


class SmallestInfiniteSet:
    def __init__(self):
        self.heap = list(range(1, 1001))  # cap initial range reasonably
        heapq.heapify(self.heap)
        self.present = set(self.heap)

    def popSmallest(self) -> int:
        if not self.heap:
            # expand if needed
            v = max(self.present) + 1 if self.present else 1
            return v
        x = heapq.heappop(self.heap)
        self.present.remove(x)
        return x

    def addBack(self, num: int) -> None:
        if num not in self.present:
            heapq.heappush(self.heap, num)
            self.present.add(num)


if __name__ == '__main__':
    s = SmallestInfiniteSet()
    print(s.popSmallest())  # 1
    print(s.popSmallest())  # 2
    s.addBack(1)
    print(s.popSmallest())  # 1
